storage-proofs/src/election_post.rs

```diff
@@ -0,0 +1,482 @@
+    let bits = Bits::new_many(list.iter());
+
+    let partial_ticket = pedersen_md_no_padding_bits(bits);
+
+    // ticket = sha256(partial_ticket)
+    let ticket = finalize_ticket(&partial_ticket);
+
+    Ok(Candidate {
+        sector_challenge_index,
+        sector_id,
+        partial_ticket,
+        ticket,
+    })
+}
+
+pub fn finalize_ticket(partial_ticket: &Fr) -> [u8; 32] {
+    let bytes = fr_into_bytes::<Bls12>(partial_ticket);
+    let ticket_hash = Sha256::digest(&bytes);
+    let mut ticket = [0u8; 32];
+    ticket.copy_from_slice(&ticket_hash[..]);
+    ticket
+}
+
+pub fn generate_sector_challenges(
+    randomness: &[u8; 32],
+    challenge_count: usize,
+    sectors: &OrderedSectorSet,
+) -> Result<Vec<SectorId>> {
+    let mut challenges = Vec::with_capacity(challenge_count);
+
+    for n in 0..challenge_count as usize {
+        let sector = generate_sector_challenge(randomness, n, sectors)?;
+        challenges.push(sector);
+    }
+
+    Ok(challenges)
+}
+
+pub fn generate_sector_challenge(
+    randomness: &[u8; 32],
+    n: usize,
+    sectors: &OrderedSectorSet,
+) -> Result<SectorId> {
+    let mut hasher = Sha256::new();
+    hasher.input(&randomness[..]);
+    hasher.input(&n.to_le_bytes()[..]);
+    let hash = hasher.result();
+
+    let sector_challenge = LittleEndian::read_u64(&hash.as_ref()[..8]);
+    let sector_index = (sector_challenge % sectors.len() as u64) as usize;
+    let sector = *sectors
+        .iter()
+        .nth(sector_index)
+        .expect("invalid challenge generated");
+
+    Ok(sector)
+}
+
+/// Generate all challenged leaf ranges for a single sector, such that the range fits into the sector.
+pub fn generate_leaf_challenges(
+    randomness: &[u8; 32],
+    sector_challenge_index: u64,
+    sector_size: u64,
+) -> Vec<u64> {
+    let mut challenges = Vec::with_capacity(POST_CHALLENGE_COUNT);
+
+    for leaf_challenge_index in 0..POST_CHALLENGE_COUNT {
+        let challenge = generate_leaf_challenge(
+            randomness,
+            sector_challenge_index,
+            leaf_challenge_index as u64,
+            sector_size,
+        );
+        challenges.push(challenge)
+    }
+
+    challenges
+}
+
+/// Generates challenge, such that the range fits into the sector.
+pub fn generate_leaf_challenge(
+    randomness: &[u8; 32],
+    sector_challenge_index: u64,
+    leaf_challenge_index: u64,
+    sector_size: u64,
+) -> u64 {
+    assert!(
+        sector_size > POST_CHALLENGED_NODES as u64 * NODE_SIZE as u64,
+        "sector size {} is too small",
+        sector_size
+    );
+
+    let mut hasher = Sha256::new();
+    hasher.input(&randomness[..]);
+    hasher.input(&sector_challenge_index.to_le_bytes()[..]);
+    hasher.input(&leaf_challenge_index.to_le_bytes()[..]);
+    let hash = hasher.result();
+
+    let leaf_challenge = LittleEndian::read_u64(&hash.as_ref()[..8]);
+
+    let sector_part = leaf_challenge % (sector_size / (POST_CHALLENGED_NODES * NODE_SIZE) as u64);
```

Comment:
Can this have a better name, please? Perhaps `challenged_range_index`.

Reply:
fixed